File: 2020/04/04.py
import sys

from typing import Callable, Dict, Iterable, List, TextIO

REQUIRED_KEYS: Dict[str, Callable[[str], bool]] = {
    'byr': lambda _: True,
    "iyr": lambda _: True,
    "eyr": lambda _: True,
    "hgt": lambda _: True,
    "hcl": lambda _: True,
    "ecl": lambda _: True,
    "pid": lambda _: True,
};

def parse_passports(stream: TextIO) -> Iterable[Dict[str, str]]:
    this_passport: Dict[str, str] = dict()
    for line in stream:
        line = line.rstrip()
        if line == '' and len(this_passport) > 0:
            yield this_passport
            this_passport = dict()
        for entry in line.split():
            k, v = entry.split(':')
            if k in this_passport:
                raise ValueError(f'duplicate key: {k}')
            this_passport[k] = v
    if len(this_passport) > 0:
        yield this_passport

def valid1(passport: Dict[str, str]) -> bool:
    return all(k in passport for k in REQUIRED_KEYS.keys())

def main(argv: List[str]) -> int:
    passports = parse_passports(sys.stdin)
    print(f'{sum(1 if valid1(p) else 0 for p in passports)} valid1')
    return 0

if __name__ == '__main__':
    sys.exit(main(sys.argv))

File: 2020/04/04.rs
use std::{
    fmt,
    io::{self, Read},
    error::Error,
    collections::HashMap,
};

const REQUIRED_KEYS: [&'static str; 7] = [
    "byr",
    "iyr",
    "eyr",
    "hgt",
    "hcl",
    "ecl",
    "pid",
    // "cid",
];

#[derive(Debug)]
struct Passport<'a>(HashMap<&'a str, &'a str>);

impl<'a> Passport<'a> {
    fn valid(&self) -> bool {
        for k in REQUIRED_KEYS.iter() {
            if !self.0.contains_key(k) {
                return false;
            }
        }
        true
    }
}

#[derive(Debug)]
enum ParseError {
    InvalidEntry(String),
    MissingValue(String),
    DuplicateKey(String),
}

impl fmt::Display for ParseError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{:?}", self)
    }
}

impl Error for ParseError { }

fn parse_passports(file: &str) -> Result<Vec<Passport>, ParseError> {
    let mut passports = Vec::new();
    let mut entries = HashMap::new();
    for line in file.lines() {
        if line == "" {
            passports.push(Passport(entries));
            entries = HashMap::new();
        }

        for entry in line.split_whitespace() {
            match entry.find(':') {
                Some(ix) => {
                    let (k, v) = entry.split_at(ix);

                    if v.len() == 1 {
                        return Err(ParseError::MissingValue(k.to_owned()));
                    }

                    if let Some(_) = entries.insert(k, &v[1..]) {
                        return Err(ParseError::DuplicateKey(k.to_owned()));
                    }
                },

                None => return Err(ParseError::InvalidEntry(entry.to_owned())),
            };
        }
    }

    if !entries.is_empty() {
        passports.push(Passport(entries));
    }

    Ok(passports)
}

fn main() -> Result<(), Box<dyn Error>> {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input)?;
    let passports = parse_passports(&input)?;
    let valid_count = passports.iter().filter(|p| p.valid()).count();
    dbg!(passports);
    println!("{} valid passports", valid_count);
    Ok(())
}
